src/v/redpanda/debug_bundle.h

```diff
@@ -0,0 +1,146 @@
+#include <seastar/util/process.hh>
+
+#include <concepts>
+#include <filesystem>
+#include <memory>
+#include <optional>
+#include <stdexcept>
+#include <system_error>
+#include <vector>
+
+static constexpr ss::shard_id debug_bundle_shard_id = 0;
+
+struct debug_bundle_credentials {
+    ss::sstring username;
+    ss::sstring password;
+    ss::sstring mechanism;
+};
+
+struct debug_bundle_params {
+    std::optional<ss::sstring> logs_since;
+    std::optional<ss::sstring> logs_until;
+    std::optional<ss::sstring> logs_size_limit;
+    std::optional<ss::sstring> metrics_interval;
+
+    debug_bundle_params()
+      : logs_since{std::nullopt}
+      , logs_until{std::nullopt}
+      , logs_size_limit{std::nullopt}
+      , metrics_interval{std::nullopt} {}
+};
+
+enum class debug_bundle_status { not_running, running };
+
+constexpr static auto default_bundle_name = []() {
+    return ss::sstring{"debug-bundle.zip"};
+};
+
+constexpr static auto undefined_host_path = []() {
+    return ss::sstring{"not-defined"};
+};
+
+class debug_bundle : public ss::peering_sharded_service<debug_bundle> {
+public:
+    struct rpk_consumer {
+        using consumption_result_type =
+          typename ss::input_stream<char>::consumption_result_type;
+        using stop_consuming_type =
+          typename consumption_result_type::stop_consuming_type;
+        using tmp_buf = stop_consuming_type::tmp_buf;
+
+        rpk_consumer(debug_bundle& b)
+          : bundle{b} {}
+
+        ss::future<consumption_result_type> operator()(tmp_buf buf);
+
+        debug_bundle& bundle;
+    };
+
+    debug_bundle(
+      const std::filesystem::path& write_dir,
+      const std::filesystem::path& rpk_path);
+
+    ss::future<> start();
+    ss::future<std::error_code> start_creating_bundle(
+      const debug_bundle_params params,
+      std::optional<debug_bundle_credentials> creds);
+    ss::future<> stop();
+    ss::future<std::error_code> remove_bundle(ss::sstring& filename);
+    ss::future<std::error_code> abort_bundle();
+
+    ss::future<debug_bundle_status> get_status();
+    const std::filesystem::path& get_write_dir() { return _write_dir; }
+    ss::future<std::error_code> is_ready(ss::sstring& filename);
+
+    template<
+      std::invocable<> Func,
+      typename Futurator = ss::futurize<std::invoke_result_t<Func>>>
+    typename Futurator::type with_lock(Func&& func) {
+        gate_guard guard{_rpk_gate};
+        auto units = co_await _rpk_mu.get_units();
+        co_return co_await Futurator::invoke(std::forward<Func>(func));
+    }
+
+private:
+    friend struct rpk_consumer;
+
+    debug_bundle_status _status;
+    const std::filesystem::path _write_dir;
+    ss::sstring _host_path;
+    const std::filesystem::path _rpk_cmd;
+    ss::gate _rpk_gate;
+    mutex _rpk_mu;
+    std::optional<ss::experimental::process> _process;
+
+    ss::future<> run_rpk(
+      const debug_bundle_params params,
+      std::optional<debug_bundle_credentials> creds);
```

Comment:
All of this may be easier (both to develop, maintain, and unit-test) if we separate the `_process` handling out of `debug-bundle`.

What I'm suggesting is wrapping the process control within a separate object that handles all of the "process" stuff and just have `debug-bundle` handle the files and creating and waiting on this process class

Something like

```c++
class process_util {
public:
process_util(vec<string> args);
ss::future<errc> run(); // <-- this will execute the process and handle wait and all that
ss::future<> abort();
private:
std::optional<ss::process> _process;
};
```

This can be the class you can thoroughly unit test for things like hung processes and handling aborts and signals and all that.

Reply:
@dotnwat probably has a better architectural idea than this